### PR TITLE
Adding Deployment Stacks cmdlet Implementations to Az.Resources

### PR DESCRIPTION
**Pull request by dantedallag**
## Description

Adding new Deployment Stack cmdlets to the Az.Resources module. These will enable Deployment Stacks Create, Update, Delete, and Retrieval through PowerShell.

Along with the implementation of these cmdlets, the checked-in code for the resource manager sdk client was updated to reflect required changes.  

Link to most recent design review: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1298
Swagger for checked-in sdk: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/resources/resource-manager/Microsoft.Resources/preview/2022-08-01-preview/deploymentStacks.json

## Checklist

- [x ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Review by vidai-msft — CHANGES_REQUESTED** [review]

**Comment by vidai-msft** [comment]
@dantedallag Except for the comments above, please also notice the followings.
- Check in the `README.md` under the `Resources.Sdk` folder
- For the help docs under the `help` folder, please fill in the DESCRIPTION and EXAMPLES sections
- There are static analysis issues detected in the CI regarding the URL of the online doc, please change them accordingly.
- The test `TestExportManagementGroupDeploymentStackTemplate` failed in the CI, please fix it.

**Comment by vidai-msft** [comment]
> Hi @vidai-msft Vincent Dai FTE - we are working on the doc fixes. Can you please confirm that the rest of the PR has been reviewed? If not, when can we expect that to be done? We are targeting the 6/26 code complete date and want to ensure we've addressed all comments in time.

@apclouds I have listed all the issues in the comment. Some stuff needs to be reviewed again once files are added and updated.

**Comment by vidai-msft** [comment]
@dantedallag As mentioned in the email, please add the attribute for all the new cmdlets `[CmdletPreview("The cmdlet is in preview and under development.")]`

File: src/Resources/Resources/help/Set-AzManagementGroupDeploymentStack.md [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+## SYNOPSIS
+Sets a new Management Group scoped Deployment Stack.
+
+## SYNTAX
+
+### ByTemplateFileWithNoParameters (Default)
+```
+Set-AzManagementGroupDeploymentStack -Name <String> -ManagementGroupId <String>
+ -DeploymentSubscriptionId <String> -Location <String> [-Description <String>] [-DeleteAll] [-DeleteResources]
+ [-DeleteResourceGroups] -DenySettingsMode <PSDenySettingsMode> [-DenySettingsExcludedPrincipals <String[]>]
```
Comment (vidai-msft) [inline src/Resources/Resources/help/Set-AzManagementGroupDeploymentStack.md:19]: Use singular form for parameters

File: src/Resources/Resources/help/Set-AzResourceGroupDeploymentStack.md [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+## SYNOPSIS
+Sets a new Resource Group scoped Deployment Stack.
+
+## SYNTAX
+
+### ByTemplateFileWithNoParameters (Default)
+```
+Set-AzResourceGroupDeploymentStack [-Name] <String> [-ResourceGroupName] <String> [-Description <String>]
+ [-DeleteAll] [-DeleteResources] [-DeleteResourceGroups] -DenySettingsMode <PSDenySettingsMode>
+ [-DenySettingsExcludedPrincipals <String[]>] [-DenySettingsExcludedActions <String[]>]
```
Comment (vidai-msft) [inline src/Resources/Resources/help/Set-AzResourceGroupDeploymentStack.md:19]: Use singular form for parameters

File: src/Resources/Resources/help/Set-AzSubscriptionDeploymentStack.md [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+Sets a new Subscription scoped Deployment Stack.
+
+## SYNTAX
+
+### ByTemplateFileWithNoParameters (Default)
+```
+Set-AzSubscriptionDeploymentStack -Name <String> [-Description <String>] -Location <String>
+ [-DeploymentResourceGroupName <String>] [-DeleteAll] [-DeleteResources] [-DeleteResourceGroups]
+ -DenySettingsMode <PSDenySettingsMode> [-DenySettingsExcludedPrincipals <String[]>]
+ [-DenySettingsExcludedActions <String[]>] [-DenySettingsApplyToChildScopes] [-Tag <Hashtable>] [-Force]
```
Comment (vidai-msft) [inline src/Resources/Resources/help/Set-AzSubscriptionDeploymentStack.md:20]: Use singular form for parameters

File: src/Resources/ResourceManager/Implementation/DeploymentStacks/ExportAzManagmentGroupDeploymentStackTemplate.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ExportByResourceIdParameterSetName, 
+            HelpMessage = "ResourceId of the DeploymentStack to get")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Alias("StackName")]
+        [Parameter(Position = 1, Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ExportByNameAndManagementGroupIdParameterSetName, 
+            HelpMessage = "The name of the DeploymentStack to get")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
```
Comment (vidai-msft) [inline src/Resources/ResourceManager/Implementation/DeploymentStacks/ExportAzManagmentGroupDeploymentStackTemplate.cs:44]: The property name should be `StackName`

File: src/Resources/ResourceManager/Implementation/DeploymentStacks/ExportAzManagmentGroupDeploymentStackTemplate.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        internal const string ExportByResourceIdParameterSetName = "ExportByResourceId";
+        internal const string ExportByNameAndManagementGroupIdParameterSetName = "ExportByNameAndManagmentGroupId";
+
+        [Alias("Id")]
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ExportByResourceIdParameterSetName, 
+            HelpMessage = "ResourceId of the DeploymentStack to get")]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Alias("StackName")]
```
Comment (vidai-msft) [inline src/Resources/ResourceManager/Implementation/DeploymentStacks/ExportAzManagmentGroupDeploymentStackTemplate.cs:40]: No alias needed

**Review by vidai-msft — CHANGES_REQUESTED** [review]

**Comment by vidai-msft** [comment]
@dantedallag looks like this PR has no associated design review. So some parameters do not comply with PowerShell design guideline. I pointed them out on the help docs. You need to change the code first and then re-generate using platyPS. In addition to that, I also have a question regarding the Set-* cmdlets. From the swagger file, seems they all use `put`. Since some parameters there are optional, if one parameter is not provided, then would its value be cleared from the server side?

**Comment by vidai-msft** [comment]
@dantedallag can you please provide the full URL of the swagger file so that I can help verify the code generation part? Thanks!

**Comment by vidai-msft** [comment]
> For the parameter changes from plural to singular requested, we will make the change for "DenySettingsExcludedPrincipals" and "DenySettingsExcludedActions". For the other, given they do not take in a list as input, we strongly feel this would cause a lot of confusion with our customers if these were changed to singular.

@apclouds As mentioned by you, since the switches `DeleteResources` and `DeleteResourceGroups` do not take a list as input, I feel like it should better use singular form. Your thoughts?

**Comment by vidai-msft** [comment]
> Could I also get some more context on what exactly is failing in the help generation? I added the examples and description sections.

@dantedallag I have already mentioned in the review comments but wanted to re-iterate it here. The issue was caused by online link shown as docs.microsoft.com. Now this URL has been updated to learn.microsoft.com

File: src/Resources/Resources/help/Export-AzManagementGroupDeploymentStackTemplate.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+# Export-AzManagementGroupDeploymentStackTemplate
+
+## SYNOPSIS
+Exports a Management Group scoped Deployment Stack Template.
+
+## SYNTAX
+
+### ExportByNameAndManagmentGroupId (Default)
+```
+Export-AzManagementGroupDeploymentStackTemplate [-Name] <String> [-ManagementGroupId] <String> [-Pre]
```
[the remark this answers is not in the document]
Comment (vidai-msft) [reply src/Resources/Resources/help/Export-AzManagementGroupDeploymentStackTemplate.md:17]: @apclouds Other cmdlets are [Verb]-[Prefix]DeploymentStack while the Export is Export-[Prefix]DeploymentStackTemplate. In the world of PowerShell, the DeploymentStack implies the name is StackName while DeploymentStackTemplate implies it should be TemplateName. This is a convention in PowerShell and an experienced PowerShell user would be confused.

**Comment by vidai-msft** [comment]
@dantedallag Can you please remove the prompt `PS C:\>` in all the examples? The static analyzer failed because of that.

**Comment by vidai-msft** [comment]
> > > For the parameter changes from plural to singular requested, we will make the change for "DenySettingsExcludedPrincipals" and "DenySettingsExcludedActions". For the other, given they do not take in a list as input, we strongly feel this would cause a lot of confusion with our customers if these were changed to singular.
> > 
> > 
> > @apclouds As mentioned by you, since the switches `DeleteResources` and `DeleteResourceGroups` do not take a list as input, I feel like it should better use singular form. Your thoughts?
> 
> Hi @vidai-msft, today the majority of Top 100 Azure customers are deploying multiple resources at a time defined in a single ARM/Bicep template. When the deployment of that template is done via Deployment Stack, the resources deployed become managed by the stack resource, and both switches of `DeleteResources` and `DeleteResourceGroups` specify the behavior of "deleting all resources" or "deleting all resources groups", rather than just deleting a single resource/resource-group. We believe this will cause confusion in the customer experience, making it seem like only a single resource will be deleted. Even though the switches do not take in lists as input, the plural format helps more clearly define the stacks update behavior intent. Would it be possible to use plural given the scenario?

Agreed. Thanks for the explanation.

**Comment by vidai-msft** [comment]
> In addition to that, I also have a question regarding the Set-* cmdlets. From the swagger file, seems they all use `put`. Since some parameters there are optional, if one parameter is not provided, then would its value be cleared from the server side?

@apclouds @dantedallag Can you please help answer this question?

**Comment by Xynoclafe** [comment]
> Also from the email sent this morning, would you update from `Export` to `Save`?
This has been updated

**Comment by vidai-msft** [comment]
> > > In addition to that, I also have a question regarding the Set-* cmdlets. From the swagger file, seems they all use `put`. Since some parameters there are optional, if one parameter is not provided, then would its value be cleared from the server side?
> > 
> > 
> > @apclouds
> >             
> >             Angel Perez
> >             FTE @dantedallag
> >             
> >             Dante Dalla Gasperina
> >             FTE Can you please help answer this question?
> 
> Yes, that is correct. Both New and Set use the same underlying Put call that overwrites the existing stack. The main difference between the two is that New will ask for a confirmation if the stack already exists, as it is expecting the stack to be new. You can however overwrite a stack with the New command by acknowledging this confirmation.

@dantedallag Say a deployment stack was created via New-* with tags associated. If later on, the user ran Set-* cmdlet to update some property but without providing the existing value of tag. As per what you have mentioned, the tag would be cleaned up. If so, I believe it would cause problem as the users may not be aware of the deletion for tag. Our proposal is to retrieve the deployment stack object first and then replace the values that have been explicitly provided by the Set-* cmdlet. Then you won't delete something unintentionally.

**Comment by vidai-msft** [comment]
> > > > > In addition to that, I also have a question regarding the Set-* cmdlets. From the swagger file, seems they all use `put`. Since some parameters there are optional, if one parameter is not provided, then would its value be cleared from the server side?
> > > > 
> > > > 
> > > > @apclouds
> > > >             
> > > >             Angel Perez
> > > >             FTE
> > > > ```
> > > >         Angel Perez
> > > >         FTE @dantedallag
> > > >         
> > > >         Dante Dalla Gasperina
> > > >         FTE Can you please help answer this question?
> > > > ```
> > > 
> > > 
> > > Yes, that is correct. Both New and Set use the same underlying Put call that overwrites the existing stack. The main difference between the two is that New will ask for a confirmation if the stack already exists, as it is expecting the stack to be new. You can however overwrite a stack with the New command by acknowledging this confirmation.
> > 
> > 
> > @dantedallag
> >             
> >             Dante Dalla Gasperina
> >             FTE Say a deployment stack was created via New-* with tags associated. If later on, the user ran Set-* cmdlet to update some property but without providing the existing value of tag. As per what you have mentioned, the tag would be cleaned up. If so, I believe it would cause problem as the users may not be aware of the deletion for tag. Our proposal is to retrieve the deployment stack object first and then replace the values that have been explicitly provided by the Set-* cmdlet. Then you won't delete something unintentionally.
> 
> We plan to keep the existing behavior and gather more customer feedback before this change. Stacks are a proxy resource and come with limitations around tags scenarios. Therefore, we don't expect significant usage on tags and don't consider this a blocking issue for Stacks. Can we please get approval on this PR without this change? We will get more customer feedback on this once released.

@apclouds It is OK. Wanted to reiterate that this can be applied to any non-mandatory parameters besides Tag. Tag is just an example easy to understand. Thanks!

**Comment by vidai-msft** [comment]
@Xynoclafe @dantedallag In the Save-*, the method names in DeploymentStackSdkClient are still `ExportManagementGroupDeploymentStack`, `ExportResourceGroupDeploymentStack`, `ExportSubscriptionDeploymentStack`. Do you want to update as well?

**Comment by vidai-msft** [comment]
> > @Xynoclafe @dantedallag In the Save-*, the method names in DeploymentStackSdkClient are still `ExportManagementGroupDeploymentStack`, `ExportResourceGroupDeploymentStack`, `ExportSubscriptionDeploymentStack`. Do you want to update as well?
> 
> @vidai-msft Yes, we can make this change. After this change, is the PR at a state where we can get approval and merge? Let us know if there are any final things we need to do besides this so that we can get approval from you tomorrow and hopefully merge during our workday on Thursday. We want to ensure that we can get this code in as soon as possible so we comfortably make the check-in deadline for the next release.

@dantedallag Yes, should have no other problems except this.

**Review by vidai-msft — APPROVED** [review]
LGTM

**Pull request merged**